feat: Add deep link and launcher shortcut support for conversations

angrymuesli opened the pull request:
- Add custom URI scheme (nctalk://conversation/{token}) for opening conversations from external launchers like KISS
- Add HTTPS deep link support for /call/{token} URLs (fixes #847)
- Add dynamic shortcuts for favorite/recent conversations
- Add "Add to home screen" menu option in conversation long-press dialog
- New DeepLinkHandler utility for parsing deep link URIs
- New ShortcutManagerHelper utility for managing conversation shortcuts

### 🖼️ Screenshots

🏚️ Before | 🏡 After
---|---
B | A


### 🚧 TODO

- [ ] ...

### 🏁 Checklist

- [ ] ⛑️ Tests (unit and/or integration) are included or not needed
- [ ] 🔖 Capability is checked or not needed 
- [ ] 🔙 Backport requests are created or not needed: `/backport to stable-xx.x`
- [ ] 📅 Milestone is set
- [ ] 🌸 PR title is meaningful (if it should be in the changelog: is it meaningful to users?)

angrymuesli committed:
feat: Add deep link and launcher shortcut support for conversations

- Add custom URI scheme (nctalk://conversation/{token}) for opening
  conversations from external launchers like KISS
- Add HTTPS deep link support for /call/{token} URLs (fixes #847)
- Add dynamic shortcuts for favorite/recent conversations
- Add "Add to home screen" menu option in conversation long-press dialog
- New DeepLinkHandler utility for parsing deep link URIs
- New ShortcutManagerHelper utility for managing conversation shortcuts

Signed-off-by: angrymuesli <github.visibly626@slmails.com>

angrymuesli commented:
> @angrymuesli - I tested this PR, shortcut support for conversations work well. Opening Talk links from different domains and launching conversations via the external KISS launcher works only on Android 12 and earlier devices but not on Android 12+ devices. Deep links only work if the domain is registered via assetlinks.json. Talk is a self-hosted and can run on many different domains, it is not feasible for us to register and verify all possible domains. As a result, automatically opening Talk links from different domains in the app on Android 12+ devices is practically not implementable.

Hi, and thanks for the feedback! Not all of that is true. I installed a debug build based on the latest changes on my Pixel 6a running GrapheneOS (Android 16), and I could easily access Talk conversations in KISS by typing a few letters of the conversation name. Did you mean something else?

angrymuesli commented:
> To resolve linter issues, please run _./gradlew ktlintFormat_
> 
> There are also some static code analysis issues, see https://app.codacy.com/gh/nextcloud/talk-android/pull-requests/5664/issues. Please feel free to fix them, or let us know if you want us to handle them.

Sure, I'll run the linter and address the static code analysis issues. Thanks for pointing these out.

sowjanyakch commented:
> Hi, and thanks for the feedback! Not all of that is true. I installed a debug build based on the latest changes on my Pixel 6a running GrapheneOS (Android 16), and I could easily access Talk conversations in KISS by typing a few letters of the conversation name. Did you mean something else?

Talk conversations could be accessed when I used the KISS launcher as default launcher on mobile, however i noticed a issue - When I try to access a conversation, sometimes a different conversation opens, sometimes the requested conversation opens and sometimes it just opens the conversations list.  Please see the below video and I used single account for testing purpose. 

https://github.com/user-attachments/assets/357cc8ad-b586-4e6a-a82f-4c0d2cff8b7e

Additionally, the other thing, that does not work is opening different domain talk links on the talk app, it redirects to web browser instead of app for Android 12+ devices.

mahibi reviewed (changes requested):
see comment from https://github.com/nextcloud/talk-android/pull/5664#issuecomment-3759835609

mahibi commented:
HI @angrymuesli 

we decided that the schema that is described in 
https://github.com/nextcloud/spreed/issues/16354
will be taken as is. 

Also, i suggest that implementing https deep links should not be done as it's already considered as legacy.

Could you implement these changes? Just let us know if there are any questions!

angrymuesli committed:
refactor: implement nextcloudtalk:// URI scheme per spreed#16354

Address maintainer feedback and improve code quality:

- Replace nctalk:// with nextcloudtalk:// custom URI scheme
- Remove HTTPS deep links (not feasible for self-hosted domains)
- Add token validation (alphanumeric, 4-32 chars)
- Fix server URL matching to use proper host comparison
- Handle null user.id gracefully in ShortcutManagerHelper
- Manage RxJava disposables to prevent memory leaks
- Add lifecycle checks before UI operations
- Fix generic exception handling

URI format: nextcloudtalk://[user@]host/[base/]call/token

Signed-off-by: angrymuesli <github.visibly626@slmails.com>

angrymuesli commented:
Hi @sowjanyakch, yes I'm continuing! Just force-pushed with:
- Rebased on latest master
- Fixed all Codacy static analysis issues (ReturnCount)
- Added DCO sign-off
- ktlint and detekt pass locally

The PR implements the `nextcloudtalk://` URI scheme as decided in spreed#16354, with no HTTPS deep links. Ready for review when you have time!

sowjanyakch reviewed (changes requested):
1.  Please align the menu icon and title with other options.

<img width="355" height="400" alt="Image" src="https://github.com/user-attachments/assets/d62d7c56-8281-4c72-87b7-3aec1051b0c4" />

2. Deleting a conversation should also remove shortcut or make the shortcut disabled and show a warning shortcut is unavailable/conversation no longer available when clicked.

3. When multiple accounts A, B are present, create a shortcut for a conversation belonging to account A and then log out of account A. If the shortcut is opened, the app correctly opens the conversation list for account B. Now log in to account A and tap the same shortcut again, the app now switches to account A but displays the conversation list of account B instead of navigating to the intended conversation.

<img width="400" height="142" alt="Image" src="https://github.com/user-attachments/assets/82c15a28-94f5-44e5-a858-bad8815df649" />

angrymuesli committed:
fix: address review feedback for shortcuts and deep links

- Align "Add to home screen" menu icon and text padding with other
  bottom dialog items (use standard_dialog_padding consistently)
- Disable shortcuts when a conversation is deleted so pinned shortcuts
  show "conversation no longer exists" instead of broken navigation
- Use nextcloudtalk:// deep link URIs in shortcuts instead of bundle
  extras to fix multi-account navigation bug where switching accounts
  via shortcut would show conversation list instead of the conversation

Signed-off-by: angrymuesli <github.visibly626@slmails.com>

angrymuesli committed:
fix: handle shortcut cleanup in ConversationInfoActivity and fix back-stack

- Disable shortcuts when deleting or leaving a conversation from
  ConversationInfoActivity (previously only handled in bottom dialog)
- Fix back-stack for deep link navigation: open ConversationsListActivity
  before ChatActivity so pressing back shows the correct user's
  conversation list instead of stale data from a previous account

Signed-off-by: angrymuesli <github.visibly626@slmails.com>

sowjanyakch commented:
> 2. **Back-stack** — Deep link navigation now uses `startActivities()` to place `ConversationsListActivity` underneath `ChatActivity`. Pressing back now correctly shows the target user's conversation list.

The current user is not properly set to the target user before navigating to the conversation list, so the issue still persists.

angrymuesli committed:
feat: adapt shortcuts to Compose conversation operations sheet

Upstream migrated the conversation bottom dialog to a Compose
ModalBottomSheet. Adapt our shortcut features to the new architecture:

- Add AddToHomeScreen action to ConversationOpsAction sealed class
- Add "Add to home screen" menu item in ConversationOperationsSheet
- Wire handler in ConversationsListActivity.handleConversationOpsAction
- Fix back-stack: pass KEY_INTERNAL_USER_ID to ConversationsListActivity
  via intent so it reads the correct user from DB instead of stale cache

Signed-off-by: angrymuesli <github.visibly626@slmails.com>

angrymuesli commented:
Hi @sowjanyakch, pushed with the fixes:

- **Back-stack fix**: ConversationsListActivity now reads the target user directly from the intent via `KEY_INTERNAL_USER_ID`, bypassing the stale provider cache. Pressing back from a shortcut-opened chat now shows the correct user's conversation list.
- **Compose migration**: Adapted "Add to home screen" to the new ModalBottomSheet architecture (`ConversationOpsAction` + `ConversationOperationsSheet`).
- **Shortcut cleanup**: All delete/leave paths now covered (ConversationsListActivity + ConversationInfoActivity).

Rebased on latest master, all checks pass locally. Looking forward to your feedback whenever you get a chance!

mahibi commented:
Thank you for keeping it up @angrymuesli we will review again soon

sowjanyakch commented:
Tested and works as expected.

Note to myself: The existing bug with the account switching also impacts this PR and it needs to be fixed.

angrymuesli commented:
> Tested and works as expected.
> 
> Note to myself: The existing bug with the account switching also impacts this PR and it needs to be fixed.

Could you share the details of this account switcher bug, maybe that could be easily fixed within this PR?

sowjanyakch commented:
> Could you share the details of this account switcher bug, maybe that could be easily fixed within this PR?

When using multi account, sometimes chats from another account are sometimes displayed in the current account. This is an existing bug and something we need to investigate.

angrymuesli commented:
Thanks @sowjanyakch — looks like this is already tracked in #4039. Happy to help investigate in a separate PR once the root cause is identified.

I've rebased on latest master to resolve the new conflicts. Since you've confirmed it works as expected, could you formally approve when you have a moment? There are some older review states from earlier feedback rounds that GitHub is still showing as blocking — your approval would let the team move forward with the merge. Thanks!

sowjanyakch committed:
parse link properly

Signed-off-by: sowjanyakch <sowjanya.kch@gmail.com>

sowjanyakch commented:
@angrymuesli - I made some changes to the PR covering some edge cases like username will be xyz or xyz@cloud.com. This PR handled only first case due to wrong parsing. I will deep test again with multiple accounts from different servers and after that it's good to merge.

sowjanyakch commented:
Notes : This PR works as expected but there is a problem with the account switching. Currently account switching does not work properly and it is observed multiple times on the latest master. This PR can be merged but can only be included in release once we fix the problem with the account switching.

mahibi commented:
i took over in https://github.com/nextcloud/talk-android/pull/6207 to rebase on master, fix conflicts and also fixed a bug with opening multiple app instances when clicking a deep link.
Hope that's okay for you @angrymuesli 
Thanks again to this great contribution :+1:
